library/src/main/java/io/github/boguszpawlowski/composecalendar/month/MonthListState.kt

```diff
@@ -5,33 +5,43 @@ import androidx.compose.runtime.Stable
 import androidx.compose.runtime.derivedStateOf
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.snapshotFlow
-import io.github.boguszpawlowski.composecalendar.header.MonthState
+import io.github.boguszpawlowski.composecalendar.selection.DynamicSelectionState
+import io.github.boguszpawlowski.composecalendar.selection.SelectionState
+import io.github.boguszpawlowski.composecalendar.states.CurrentState
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.launchIn
 import kotlinx.coroutines.flow.onEach
 import kotlinx.coroutines.launch
+import java.time.LocalDate
 import java.time.YearMonth
 import java.time.temporal.ChronoUnit
 
 @Stable
 internal class MonthListState(
   private val coroutineScope: CoroutineScope,
   private val initialMonth: YearMonth,
-  private val monthState: MonthState,
+  private val currentState: CurrentState,
   private val listState: LazyListState,
+  private val selectionState: DynamicSelectionState?
 ) {
 
   private val currentlyVisibleMonth by derivedStateOf {
     getMonthForPage(listState.firstVisibleItemIndex)
   }
 
   init {
-    snapshotFlow { monthState.currentMonth }.onEach { month ->
+    snapshotFlow { YearMonth.of(
+      currentState.day.year,
+      currentState.day.month
+    ) }.onEach { month ->
       moveToMonth(month)
     }.launchIn(coroutineScope)
 
     snapshotFlow { currentlyVisibleMonth }.onEach { newMonth ->
-      monthState.currentMonth = newMonth
+      currentState.day = if (selectionState != null && selectionState.selection.isNotEmpty())
+        if (selectionState.selection[0].month == newMonth.month) newMonth.atDay(selectionState.selection[0].dayOfMonth)
+          else newMonth.atDay(1)
+      else newMonth.atDay(1)
```

Comment:
I think this logic is too complicated to be handled here. We should extract it to another place.